src/rgw/s3select.h

```diff
@@ -0,0 +1,852 @@
+
+/// AST builder
+
+class s3select_projections {
+
+    private:
+        list<base_statement *> m_projections;
+    
+    public:
+        bool is_aggregate()
+        {
+            //TODO iterate on projections , and search for aggregate
+            //for(auto p : m_projections){}
+
+	    return false;
+        }
+
+        bool semantic()
+        {
+            //TODO check aggragtion function are not nested 
+            return false;
+        }
+
+        list<base_statement *> * get()
+        {
+            return &m_projections;
+        }
+
+};
+
+struct actionQ
+{
+
+    list<mulldiv_operation::muldiv_t> muldivQ;
+    list<addsub_operation::addsub_op_t> addsubQ;
+    list<arithmetic_operand::cmp_t> arithmetic_compareQ;
+    list<logical_operand::oplog_t> logical_compareQ;
+    list<base_statement *> exprQ;
+    list<base_statement *> funcQ;
+    list<base_statement *> condQ;
+    std::string from_clause;
+    list<std::string> schema_columns;
+    s3select_projections  projections;
+
+};
+
+class base_action
+{
+
+public:
+    base_action() : m_action(0) {}
+    actionQ *m_action;
+    void set_action_q(actionQ *a) { m_action = a; }
+};
+
+struct push_from_clause : public base_action
+{
+    void operator()(const char *a, const char *b) const
+    {
+        std::string token(a, b);
+        m_action->from_clause = token;
+    }
+
+};
+static push_from_clause g_push_from_clause;
+
+struct push_number : public base_action //TODO use define for defintion of actions
+{
+    void operator()(const char *a, const char *b) const
+    {
+        string token(a, b);
+        variable *v = new variable(atoi(token.c_str())); //TODO strtoll
+
+        m_action->exprQ.push_back(v);
+    }
+
+};
+static push_number g_push_number;
+
+struct push_float_number : public base_action //TODO use define for defintion of actions
+{
+
+    void operator()(const char *a, const char *b) const
+    {
+        string token(a, b);
+
+        //the parser for float(real_p) is accepting also integers, thus "blocking" integer acceptence and all are float.
+        parse_info<> info = parse(token.c_str(), int_p, space_p);
+
+        if (!info.full)
+        {
+            char *perr;
+            double d = strtod(token.c_str(), &perr);
+            variable *v = new variable(d);
+
+            m_action->exprQ.push_back(v);
+        }
+        else
+        {
+            variable *v = new variable(atoi(token.c_str())); //TODO strtoll
+
+            m_action->exprQ.push_back(v);
+        }
+    }
+
+};
+static push_float_number g_push_float_number;
+
+struct push_string : public base_action //TODO use define for defintion of actions
+{
+
+    void operator()(const char *a, const char *b) const
+    {
+        a++;b--;// remove double quotes
+        string token(a, b);
+        
+        //strdup allocation should release only at the end of query-execution (class{runtime_allocator} not accessing those allocationr) 
```

Comment:
ahem.  appearance of strdup in the comment reinforces my sense that maybe the code is more C and less c++ just out of familiarity

Reply:
will be fixed